test/test_samplers.py

```diff
@@ -0,0 +1,191 @@
+            print('sampled_vars: {}'.format(', '.join(sampled_vars)))
+            try:
+                y = x.sample(sampled_vars, sample_inputs)
+            except NotImplementedError:
+                xfail = True
+                continue
+            if sampled_vars:
+                assert dict(y.inputs) == dict(expected_inputs), sampled_vars
+            else:
+                assert y is x
+    if xfail:
+        pytest.xfail(reason='Not implemented')
+
+
+@pytest.mark.parametrize('sample_inputs', [
+    (),
+    (('s', bint(6)),),
+    (('s', bint(6)), ('t', bint(7))),
+], ids=id_from_inputs)
+@pytest.mark.parametrize('int_event_inputs', [
+    (),
+    (('d', bint(2)),),
+    (('d', bint(2)), ('e', bint(3))),
+], ids=id_from_inputs)
+@pytest.mark.parametrize('real_event_inputs', [
+    (('g', reals()),),
+    (('g', reals()), ('h', reals(4))),
+], ids=id_from_inputs)
+def test_joint_shape(sample_inputs, int_event_inputs, real_event_inputs):
+    event_inputs = int_event_inputs + real_event_inputs
+    discrete_inputs = OrderedDict(int_event_inputs)
+    gaussian_inputs = OrderedDict(event_inputs)
+    expected_inputs = OrderedDict(sample_inputs + event_inputs)
+    sample_inputs = OrderedDict(sample_inputs)
+    event_inputs = OrderedDict(event_inputs)
+    t = random_tensor(discrete_inputs)
+    g = random_gaussian(gaussian_inputs)
+    x = Joint(discrete=t, gaussian=g)
+
+    xfail = False
+    for num_sampled in range(len(event_inputs)):
+        for sampled_vars in itertools.combinations(list(event_inputs), num_sampled):
+            sampled_vars = frozenset(sampled_vars)
+            print('sampled_vars: {}'.format(', '.join(sampled_vars)))
+            try:
+                y = x.sample(sampled_vars, sample_inputs)
+            except NotImplementedError:
+                xfail = True
+                continue
+            if sampled_vars:
+                assert dict(y.inputs) == dict(expected_inputs), sampled_vars
+            else:
+                assert y is x
+    if xfail:
+        pytest.xfail(reason='Not implemented')
+
+
+@pytest.mark.parametrize('batch_inputs', [
+    (),
+    (('b', bint(4)),),
+    (('b', bint(4)), ('c', bint(2))),
+], ids=id_from_inputs)
+@pytest.mark.parametrize('event_inputs', [
+    (('e', bint(2)),),
+    (('e', bint(2)), ('f', bint(3))),
+], ids=id_from_inputs)
+def test_tensor_distribution(event_inputs, batch_inputs):
+    num_samples = 50000
+    sample_inputs = OrderedDict(n=bint(num_samples))
+    be_inputs = OrderedDict(batch_inputs + event_inputs)
+    batch_inputs = OrderedDict(batch_inputs)
+    event_inputs = OrderedDict(event_inputs)
+    sampled_vars = frozenset(event_inputs)
+    p = random_tensor(be_inputs)
+
+    q = p.sample(sampled_vars, sample_inputs) - ops.log(num_samples)
+    mq = materialize(q).reduce(ops.logaddexp, 'n')
+    mq = mq.align(tuple(p.inputs))
+    assert_close(mq, p, atol=0.1, rtol=None)
+
+
+@pytest.mark.skip(reason='infinite loop')
```

Comment:
@eb8680 Could we chat about this test, which is the final task of this PR. The test exercises the `(p.exp() * f).sum()` operation we discussed yesterday, but I haven't figured out how to eagerly execute that correctly.

Reply:
Yeah let's go through this tomorrow